### ISSUE TITLE
Using instanceWeights in the DAG

### ISSUE THREAD

**Issue by Nicholas-Schaub** [vlfeat/matconvnet#943]
I am trying to apply a weight to every label I have since I am trying to run an L2 loss function on data with a range of 100-1000. The problem is that I have far fewer data points with certain values in this range (at the higher end of the range), and I think this is causing the regression to more or less ignore the outliers. What I am doing now is normalizing the data where each data label (L) is normalized by (L-mean(L))^2/var(L). The problem is that I'm not sure how to send this weight to the loss function. Ideally you would send the weights to the loss function through the get_batch function, but the train_dag.m (I'm using a DAG model) doesn't appear to allow you to do this. In fact, it doesn't appear as though train_dag.m allows you to pass any additional arguments to eval function.

So, my current solution was to rewrite train_dag.m so that if two cells are returned from get_batch, then the second cell is passed to the eval function. Is there a different way of doing this?

**Comment by AruniRC** [comment]
@Nicholas-Schaub this is indeed a problem, which I am also facing right now -- re-balancing class weights in semantic segmentation. 
The values of  `InstanceWeight` are apparently fixed at the time of defining the loss layer during initialization of the network architecture. So it is not easy to pass these instance weights dynamically as input during network training.

My solution was to write an element-wise multiplication layer which takes in two inputs: prediction scores (softmax layer output) and a variable called 'ClassWeights', the same size as the output of my softmax layer. After the elemet-wise mult, I had a loss layer.

**Comment by Nicholas-Schaub** [comment]
@AruniRC I actually found a solution to this, but I don't think it would resolve the issue entirely. My solution was to add an `instance_weights` variable to the network so that my `get_batch` function would return something like `{'input',images,'label',labels,'instance_weights',weights}`. Since I wrote my own regression function, I just added 'instance_weights' as a parameter so that if that 'instance_weights' was defined it used them, and if it wasn't defined it just set all weights to 1. I found something like this when they tried doing L1 regression in the 'fast_rcnn' example. However, I don't think this idea was incorporated into the dag loss objects. Maybe that's an update that can be made?

**Comment by Nicholas-Schaub** [comment]
Also, 'vl_nnloss' does have a way of adding instance weights. The problem is that the associated dag object doesn't allow you to set it.

**Comment by AruniRC** [comment]
@Nicholas-Schaub yes, that's similar to what I did.
The strange thing is, `vl_nnloss` has a property called `InstanceWeights` which can be set when defining the layer. However, after the network has been initialized, there does not seem to be any way of changing it during evaluation time, since it is a parameter, not a variable!

**Comment by mohsalvi** [comment]
If this is still of interest:
Define the 'classWeights' variable to be your per-class weights and set these when creating the layer. Then edit the vl_nnloss function to set instanceWeights for each instance based on the weight assigned to its class.

**Comment by agi-templar** [comment]
> If this is still of interest:
> Define the 'classWeights' variable to be your per-class weights and set these when creating the layer. Then edit the vl_nnloss function to set instanceWeights for each instance based on the weight assigned to its class.

Can you provide a sample? Not very clear about how to edit the `vl_nnloss` function. Thank you!
